gulpfile.js

```diff
@@ -124,28 +172,197 @@ gulp.task("default", function (callback) {
+    if (android && ios) emulatorTaskNameSuffix = "";
+    
+    return emulatorTaskNameSuffix
+}
 
-gulp.task("test", function (callback) {
-    runSequence("default", "test-android", "test-ios", callback);
-});
\ No newline at end of file
+var emulatorTaskNamePrefix = "emulator";
+for (var android = 0; android < 2; android++) {
+    for (var ios = 0; ios < 4; ios++) {
+        generateEmulatorTasks(emulatorTaskNamePrefix + getEmulatorTaskNameSuffix(android, ios), android, ios);
+    }
+}
+                
+function generateTestTasks(taskName, options) {
+    gulp.task(taskName + "-fast", function (callback) {
+        runTests(callback, options);
+    });
+    
+    var emulatorTaskName = emulatorTaskNamePrefix + getEmulatorTaskNameSuffix(options.android, options.ios);
+    
+    gulp.task(taskName, function (callback) {
+        runSequence("default", emulatorTaskName, taskName + fastSuffix, callback);
+    });
+    
+    gulp.task(taskName + "-clean", function (callback) {
+        runSequence("default", emulatorTaskName + cleanSuffix, taskName + fastSuffix, callback);
+    });
+}
+
+// procedurally generate tasks for every possible testing configuration
+var taskNamePrefix = "test";
+for (var android = 0; android < 2; android++) {
+    // 0 = don't run android tests
+    // 1 = run android tests
+    for (var ios = 0; ios < 4; ios++) {
+        // 0 = don't run iOS tests
+        // 1 = run iOS tests on UIWebView
+        // 2 = run iOS tests on WKWebView
+        // 3 = run iOS tests on both WebViews
+        
+        // must have at least one platform to be a test
+        if (!android && !ios) continue;
+        
+        for (var core = 0; core < 2; core++) {
+            // 0 = run all tests
+            // 1 = run only core tests
+            for (var npm = 0; npm < 2; npm++) {
+                // 0 = run tests on local version of plugin
+                // 1 = run tests on version of plugin from npm
+                
+                var taskName = taskNamePrefix;
+                var taskNameSuffix = "";
+                if (android) taskNameSuffix += "-android";
+                if (ios) taskNameSuffix += "-ios";
+                if (ios === 1) taskNameSuffix += "-uiwebview";
+                if (ios === 2) taskNameSuffix += "-wkwebview";
+                
+                // "test" instead of "test-android-ios"
+                if (android && ios === 3) taskNameSuffix = "";
+                
+                if (core) taskNameSuffix += "-core";
+                if (npm) taskNameSuffix += "-npm";
+                
+                taskName += taskNameSuffix;
+                
+                var options = {};
+                if (android) options.android = true;
+                if (ios) options.ios = true;
+                if (ios % 2 === 1) options.uiwebview = true;
+                if (ios >= 2) options.wkwebview = true;
+                if (core) options.core = true;
+                if (npm) options.npm = true;
```

Comment:
If you changed it to the following it would be cleaner IMO:

```javascript
var taskName = taskNamePrefix;
var options = {};

if (android) {
    taskNameSuffix += "-android";
    options.android = true;
}

if (ios) {
    taskNameSuffix += "-ios"
    options.ios = true;
    options.uiwebview = ios % 2 == 1;
    options.wkwebview = ios >= 2;
    if (options.uiwebview && !options.wkwebview) {
        taskNameSuffix += "-uiwebview";
    } else if (options.wkwebview && !options.uiwebview) {
        taskNameSuffix += "-wkwebview";
    }
}

if (android && ios ===3) taskNameSuffix = "";

if (core) {
    taskNameSuffix += "-core";
    options.core = true;
}

if (npm) {
    taskNameSuffix += "-npm";
    options.npm = true;
}

generateTestTasks(taskName, options);
```